# Preparing data

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from collections import Counter

Merged df has 278 rows and 124 columns. It has columns like ID, stage, gender, date, condition, VVR_group, etc. 

In [2]:
merged_df = pd.read_csv('/Users/dionnespaltman/Desktop/V6/merged_df_11-06-2024.csv', sep=',')
merged_df.drop('Unnamed: 0', axis=1, inplace=True)

display(merged_df)

ID  Stage  Gender        Date  Location  Condition  Faintness  \
0      7    1.0       2  2019-10-28         0          2        1.0   
1     23    1.0       2  2020-08-01         1          2        1.0   
2     24    1.0       2  2020-01-22         1          2        1.0   
3     25    1.0       2  2020-05-02         1          2        1.0   
4     26    1.0       1  2020-06-02         2          1        2.0   
..   ...    ...     ...         ...       ...        ...        ...   
273  323    1.0       2  2022-10-26         3          3        1.0   
274  325    1.0       2  2022-07-11         3          3        1.0   
275  329    1.0       2  2022-11-28         3          3        1.0   
276  330    1.0       2  2022-11-30         0          3        1.0   
277  331    1.0       2  2022-11-30         0          3        1.0   

     Dizziness  Weakness  Lightheadedness  ...  AU26_r__standard_deviation  \
0          1.0       1.0              2.0  ...                   -0.585935   
1          2.0       2.0              1.0  ...                   -0.624259   
2          1.0       1.0              1.0  ...                    2.367560   
3          1.0       1.0              1.0  ...                    0.584439   
4          2.0       2.0              1.0  ...                   -0.517633   
..         ...       ...              ...  ...                         ...   
273        1.0       1.0              1.0  ...                   -0.804403   
274        1.0       1.0              1.0  ...                   -1.926084   
275        1.0       1.0              1.0  ...                   -0.266200   
276        1.0       1.0              1.0  ...                   -1.889172   
277        1.0       1.0              1.0  ...                    1.340280   

     AU26_r__maximum  AU26_r__mean  AU26_r__root_mean_square  \
0           0.425279     -0.786613                 -0.696839   
1           0.425279     -0.642748                 -0.651193   
2           0.425279      3.132948                  2.795771   
3           0.425279      0.394186                  0.499951   
4           0.425279     -0.951628                 -0.731221   
..               ...           ...                       ...   
273        -0.162826     -0.846163                 -0.845794   
274         0.425279     -1.669019                 -1.846330   
275         3.521479      0.610696                  0.169932   
276        -3.328215     -1.706598                 -1.845208   
277         0.425279     -1.779581                  0.208143   

     AU45_r__sum_values  AU45_r__variance  AU45_r__standard_deviation  \
0              0.520966          1.700927                    1.712210   
1             -0.504468          0.296375                    0.440450   
2              0.129283         -0.279627                   -0.201661   
3             -0.587993          0.094357                    0.226315   
4             -1.239117         -0.246984                   -0.162414   
..                  ...               ...                         ...   
273           -0.142084          0.642278                    0.784992   
274           -0.586609         -0.362405                   -0.303044   
275           -2.381393          4.280036                    3.507346   
276           -0.958863         -0.594948                   -0.603916   
277            0.696288          6.381711                    4.701091   

     AU45_r__maximum  AU45_r__mean  AU45_r__root_mean_square  
0           0.830151      1.741972                  1.745418  
1           0.760722      1.243173                  0.665506  
2           0.014352     -0.478731                 -0.279724  
3          -0.645230      0.514820                  0.302852  
4           0.639220     -1.034420                 -0.377156  
..               ...           ...                       ...  
273         0.864866      0.560518                  0.737070  
274         0.326786     -0.488608                 -0.360082  
275

In [3]:
print(list(merged_df.columns))

['ID', 'Stage', 'Gender', 'Date', 'Location', 'Condition', 'Faintness', 'Dizziness', 'Weakness', 'Lightheadedness', 'Fear', 'Tension', 'Stress', 'Physical_sum', 'Psychological_sum', 'Sum_VVR', 'Sum_12', 'Sum_456', 'Sum_4567', 'Sum_1', 'Sum_2', 'VVR_group', 'AU01_r__sum_values', 'AU01_r__variance', 'AU01_r__standard_deviation', 'AU01_r__maximum', 'AU01_r__mean', 'AU01_r__root_mean_square', 'AU02_r__sum_values', 'AU02_r__variance', 'AU02_r__standard_deviation', 'AU02_r__maximum', 'AU02_r__mean', 'AU02_r__root_mean_square', 'AU04_r__sum_values', 'AU04_r__variance', 'AU04_r__standard_deviation', 'AU04_r__maximum', 'AU04_r__mean', 'AU04_r__root_mean_square', 'AU05_r__sum_values', 'AU05_r__variance', 'AU05_r__standard_deviation', 'AU05_r__maximum', 'AU05_r__mean', 'AU05_r__root_mean_square', 'AU06_r__sum_values', 'AU06_r__variance', 'AU06_r__standard_deviation', 'AU06_r__maximum', 'AU06_r__mean', 'AU06_r__root_mean_square', 'AU07_r__sum_values', 'AU07_r__variance', 'AU07_r__standard_deviatio

In [4]:
# Make this copy to save the information easily without having to import it again 
copy_merged_df = merged_df
print(copy_merged_df.shape)

(278, 124)


In [5]:
# columns_to_drop = ['Stage', 'Gender', 'Date', 'Location', 'Condition', 'Faintness',
#                     'Dizziness', 'Weakness', 'Lightheadedness', 'Fear', 'Tension', 'Stress', 'Physical_sum', 
#                     'Psychological_sum', 'Sum_VVR', 'Sum_12', 'Sum_456', 'Sum_4567']

columns_to_drop = ['Stage', 'Location', 'Faintness',
                    'Dizziness', 'Weakness', 'Lightheadedness', 'Fear', 'Tension', 'Stress', 'Physical_sum', 
                    'Psychological_sum', 'Sum_VVR', 'Sum_12']

merged_df = merged_df.drop(columns_to_drop, axis=1)

In [6]:
print(list(merged_df.columns))

['ID', 'Gender', 'Date', 'Condition', 'Sum_456', 'Sum_4567', 'Sum_1', 'Sum_2', 'VVR_group', 'AU01_r__sum_values', 'AU01_r__variance', 'AU01_r__standard_deviation', 'AU01_r__maximum', 'AU01_r__mean', 'AU01_r__root_mean_square', 'AU02_r__sum_values', 'AU02_r__variance', 'AU02_r__standard_deviation', 'AU02_r__maximum', 'AU02_r__mean', 'AU02_r__root_mean_square', 'AU04_r__sum_values', 'AU04_r__variance', 'AU04_r__standard_deviation', 'AU04_r__maximum', 'AU04_r__mean', 'AU04_r__root_mean_square', 'AU05_r__sum_values', 'AU05_r__variance', 'AU05_r__standard_deviation', 'AU05_r__maximum', 'AU05_r__mean', 'AU05_r__root_mean_square', 'AU06_r__sum_values', 'AU06_r__variance', 'AU06_r__standard_deviation', 'AU06_r__maximum', 'AU06_r__mean', 'AU06_r__root_mean_square', 'AU07_r__sum_values', 'AU07_r__variance', 'AU07_r__standard_deviation', 'AU07_r__maximum', 'AU07_r__mean', 'AU07_r__root_mean_square', 'AU09_r__sum_values', 'AU09_r__variance', 'AU09_r__standard_deviation', 'AU09_r__maximum', 'AU09_r

In [7]:
print(merged_df.shape)

(278, 111)


These are the columns I use to predict, so all my features. I need these as a list to establish my featurizer. 
I have 102 features from TS fresh and then I added the two VVR measurements from stage 1 and 2. 

In [15]:
columns_ts_fresh = ['AU01_r__sum_values', 'AU01_r__variance', 'AU01_r__standard_deviation', 'AU01_r__maximum', 
                    'AU01_r__mean', 'AU01_r__root_mean_square', 'AU02_r__sum_values', 'AU02_r__variance', 
                    'AU02_r__standard_deviation', 'AU02_r__maximum', 'AU02_r__mean', 'AU02_r__root_mean_square', 
                    'AU04_r__sum_values', 'AU04_r__variance', 'AU04_r__standard_deviation', 'AU04_r__maximum', 
                    'AU04_r__mean', 'AU04_r__root_mean_square', 'AU05_r__sum_values', 'AU05_r__variance',
                    'AU05_r__standard_deviation', 'AU05_r__maximum', 'AU05_r__mean', 'AU05_r__root_mean_square', 
                    'AU06_r__sum_values', 'AU06_r__variance', 'AU06_r__standard_deviation', 'AU06_r__maximum', 
                    'AU06_r__mean', 'AU06_r__root_mean_square', 'AU07_r__sum_values', 'AU07_r__variance', 
                    'AU07_r__standard_deviation', 'AU07_r__maximum', 'AU07_r__mean', 'AU07_r__root_mean_square', 
                    'AU09_r__sum_values', 'AU09_r__variance', 'AU09_r__standard_deviation', 'AU09_r__maximum',
                    'AU09_r__mean', 'AU09_r__root_mean_square', 'AU10_r__sum_values', 'AU10_r__variance', 
                    'AU10_r__standard_deviation', 'AU10_r__maximum', 'AU10_r__mean', 'AU10_r__root_mean_square',
                    'AU12_r__sum_values', 'AU12_r__variance', 'AU12_r__standard_deviation', 'AU12_r__maximum', 
                    'AU12_r__mean', 'AU12_r__root_mean_square', 'AU14_r__sum_values', 'AU14_r__variance', 
                    'AU14_r__standard_deviation', 'AU14_r__maximum', 'AU14_r__mean', 'AU14_r__root_mean_square', 
                    'AU15_r__sum_values', 'AU15_r__variance', 'AU15_r__standard_deviation', 'AU15_r__maximum', 
                    'AU15_r__mean', 'AU15_r__root_mean_square', 'AU17_r__sum_values', 'AU17_r__variance', 
                    'AU17_r__standard_deviation', 'AU17_r__maximum', 'AU17_r__mean', 'AU17_r__root_mean_square', 
                    'AU20_r__sum_values', 'AU20_r__variance', 'AU20_r__standard_deviation', 'AU20_r__maximum', 
                    'AU20_r__mean', 'AU20_r__root_mean_square', 'AU23_r__sum_values', 'AU23_r__variance', 
                    'AU23_r__standard_deviation', 'AU23_r__maximum', 'AU23_r__mean', 'AU23_r__root_mean_square', 
                    'AU25_r__sum_values', 'AU25_r__variance', 'AU25_r__standard_deviation', 'AU25_r__maximum', 
                    'AU25_r__mean', 'AU25_r__root_mean_square', 'AU26_r__sum_values', 'AU26_r__variance', 
                    'AU26_r__standard_deviation', 'AU26_r__maximum', 'AU26_r__mean', 'AU26_r__root_mean_square', 
                    'AU45_r__sum_values', 'AU45_r__variance', 'AU45_r__standard_deviation', 'AU45_r__maximum', 
                    'AU45_r__mean', 'AU45_r__root_mean_square']

print(len(columns_ts_fresh))

102


# Train and test split
May  30, 2024: 
- In total, we have 104 participants.
- I started with a test size of 20%. Then there are 83 people in the train set and 21 in the test set. With a test size of 30#, there are 72 people in the train set and 32 in the test set. 
- Naturally, we stratify on VVR_group. 

June 4, 2024: 
- In total, we have 111 participants.
- I started with a test size of 20%. Then there are 88 people in the train set and 23 in the test set. With a test size of 30% there are 77 people in the train set and 34 in the test set. 
- Naturally, we stratify on VVR_group. 

June 11, 2024: 
- In total, we have 278 participants. 
- Test size of 20%: 222 in the train and 56 in the test set. 
- Test size of 30%: 184 in the train and 84 in the test set. 

May 30, 2024: Unfortunately, the test set is very small with only 7 people in the high VVR condition. 

June 3, 2024: The test set is very small with only 11 people in the high VVR condition. 

June 11, 2024: 56 people is the low VVR group and 28 in the high VVR group. 

# Applying SMOTE
We apply SMOTE on the train data. The strategy is to make the minority as big as the majority class. 

In [16]:
from collections import Counter
from sklearn.datasets import make_classification
from imblearn.over_sampling import SMOTE

# Split data into features and target
X = merged_df.drop(columns=['ID', 'Gender', 'VVR_group', 'Date']) 
y = merged_df['VVR_group'] 

# Split data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Apply SMOTE to the training set
smote = SMOTE(sampling_strategy='not majority', random_state=42, k_neighbors=5)
X_train_smote, y_train_smote = smote.fit_resample(X_train, y_train)

print(X.shape)
print(y.shape)
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)
print(X_train_smote.shape)
print(y_train_smote.shape)


(278, 107)
(278,)
(194, 107)
(194,)
(84, 107)
(84,)
(258, 107)
(258,)


In [8]:
from collections import Counter
from sklearn.datasets import make_classification
from imblearn.over_sampling import SMOTE

# Split data into features and target
X_gender = merged_df.drop(columns=['ID', 'VVR_group', 'Date']) 
y_gender = merged_df['VVR_group'] 

# Split data into train and test sets
X_train_gender, X_test_gender, y_train_gender, y_test_gender = train_test_split(X_gender, y_gender, test_size=0.3, random_state=42)

print(X_test_gender.shape)

(84, 108)


In [9]:
X_test_gender.to_csv('/Users/dionnespaltman/Desktop/V6/X_test_gender_15-06-2024.csv')

In [17]:
X_test.to_csv('/Users/dionnespaltman/Desktop/V6/X_test_13-06-2024.csv')

In [100]:
# Print original class distribution
print('Original dataset shape %s' % Counter(y_train))

# Print resampled class distribution
print('Resampled dataset shape %s' % Counter(y_train_smote))

Original dataset shape Counter({0: 129, 1: 65})
Resampled dataset shape Counter({1: 129, 0: 129})


In [98]:
print(list(X_train.columns))

['Condition', 'Sum_456', 'Sum_1', 'Sum_2', 'AU01_r__sum_values', 'AU01_r__variance', 'AU01_r__standard_deviation', 'AU01_r__maximum', 'AU01_r__mean', 'AU01_r__root_mean_square', 'AU02_r__sum_values', 'AU02_r__variance', 'AU02_r__standard_deviation', 'AU02_r__maximum', 'AU02_r__mean', 'AU02_r__root_mean_square', 'AU04_r__sum_values', 'AU04_r__variance', 'AU04_r__standard_deviation', 'AU04_r__maximum', 'AU04_r__mean', 'AU04_r__root_mean_square', 'AU05_r__sum_values', 'AU05_r__variance', 'AU05_r__standard_deviation', 'AU05_r__maximum', 'AU05_r__mean', 'AU05_r__root_mean_square', 'AU06_r__sum_values', 'AU06_r__variance', 'AU06_r__standard_deviation', 'AU06_r__maximum', 'AU06_r__mean', 'AU06_r__root_mean_square', 'AU07_r__sum_values', 'AU07_r__variance', 'AU07_r__standard_deviation', 'AU07_r__maximum', 'AU07_r__mean', 'AU07_r__root_mean_square', 'AU09_r__sum_values', 'AU09_r__variance', 'AU09_r__standard_deviation', 'AU09_r__maximum', 'AU09_r__mean', 'AU09_r__root_mean_square', 'AU10_r__sum

In [101]:
X_train_smote.to_csv('/Users/dionnespaltman/Desktop/V6/X_train_smote_12-06-2024.csv')
y_train_smote.to_csv('/Users/dionnespaltman/Desktop/V6/y_train_smote_12-06-2024.csv')

In [103]:
X_test.to_csv('/Users/dionnespaltman/Desktop/V6/X_test_12-06-2024.csv')
y_test.to_csv('/Users/dionnespaltman/Desktop/V6/y_test_12-06-2024.csv')

# Adding class weights

I will add class weights to my models, because of my inbalanced data set. 
https://medium.com/@ravi.abhinav4/improving-class-imbalance-with-class-weights-in-machine-learning-af072fdd4aa4 

May 30, 2024: {0: 0.6428571428571429, 1: 2.25}

June 3, 2024: {0: 0.7549019607843137, 1: 1.4807692307692308}

June 11, 2024:  {0: 0.7461538461538462, 1: 1.515625}

In [102]:
import numpy as np

def calculate_class_weights(y):
    unique_classes, class_counts = np.unique(y, return_counts=True)
    total_samples = len(y)
    class_weights = {}

    for class_label, class_count in zip(unique_classes, class_counts):
        class_weight = total_samples / (2.0 * class_count)
        class_weights[class_label] = class_weight

    return class_weights

# Assuming 'y' contains the class labels (0s and 1s) for the binary classification problem
class_weights = calculate_class_weights(y_train)
print("Class weights:", class_weights)


Class weights: {0: 0.751937984496124, 1: 1.4923076923076923}


# Data distibution: old data vs new data 

In [4]:
X_train_june = pd.read_csv('/Users/dionnespaltman/Desktop/V6/data_04-06-2024/X_train_resampled.csv')
y_train_june = pd.read_csv('/Users/dionnespaltman/Desktop/V6/data_04-06-2024/y_train_resampled.csv')
X_test_june = pd.read_csv('/Users/dionnespaltman/Desktop/V6/data_04-06-2024/X_test.csv')
y_test_june = pd.read_csv('/Users/dionnespaltman/Desktop/V6/data_04-06-2024/y_test.csv')

In [5]:
X_train_may = pd.read_csv('/Users/dionnespaltman/Desktop/V4/data/X_train_res.csv')
y_train_may = pd.read_csv('/Users/dionnespaltman/Desktop/V4/data/y_train_res.csv')
X_test_may = pd.read_csv('/Users/dionnespaltman/Desktop/V4/data/X_test.csv')
y_test_may = pd.read_csv('/Users/dionnespaltman/Desktop/V4/data/y_test.csv')

In [8]:
counts_train_may = y_train_may['VVR_group'].value_counts()
print(counts_train_may)

1    56
0    56
Name: VVR_group, dtype: int64


In [9]:
counts_train_june = y_train_june['VVR_group'].value_counts()
print(counts_train_june)

1    51
0    51
Name: VVR_group, dtype: int64


In [6]:
counts_test_may = y_test_may['VVR_group'].value_counts()
print(counts_test_may)


0    25
1     7
Name: VVR_group, dtype: int64


In [7]:
counts_test_june = y_test_june['VVR_group'].value_counts()
print(counts_test_june)

0    23
1    11
Name: VVR_group, dtype: int64
